### ISSUE TITLE
StackOverflow regression with Async.TaskAwait in a recursive loop in a FSharp.Core 5.0.2

### ISSUE THREAD

**Issue by Szer** [dotnet/fsharp#11797]
# Repro steps

Provide the steps required to reproduce the problem:

```fsharp
open System
open System.Threading.Tasks

[<EntryPoint>]
let main _ =

    let rec loop (x: int) = async {
        do! Task.CompletedTask |> Async.AwaitTask
        Console.WriteLine x
        return! loop(x+1)
    }
    
    Async.RunSynchronously (loop 0)
    0
```
with
```xml
<PackageReference Update="FSharp.Core" Version="5.0.2"/>
```

`dotnet run -c Release` (both Release and Debug are affected)

# Expected behavior

It should work

# Actual behavior

Stack Overflow!

# Known workarounds

1. You could downgrade to 5.0.0 (Implicit in net5 SDK) or to 5.0.1 to make it work again:
```xml
<PackageReference Update="FSharp.Core" Version="5.0.0"/>
```
2. OR Change Completed task to fake async with real continuation
```fsharp
do! Task.CompletedTask |> Async.AwaitTask
```
to
```fsharp
do! Task.Delay 1 |> Async.AwaitTask
```

# Related information

Call stack shows that all *COMPLETED* tasks continuations started its trampoline on a stack of previous trampoline which causes SO.

It actually crashed production system which was working on [Pulsar.Client.Dotnet](https://github.com/fsharplang-ru/pulsar-client-dotnet/) (it currently using [5.0.2](https://github.com/fsharplang-ru/pulsar-client-dotnet/blob/3e1eaf1d0d7bf7f1b4c2ba5b5f7a4576956bd469/paket.lock#L15))

I also tried to find related changes with GitHub help:
https://github.com/dotnet/fsharp/compare/5.0.0..5.0.2

But I couldn't find any related to FSharp.Core tags.

* macOs
* .NET5

UPDATED: 5.0.1 also works ok. So regression was introduced between 5.0.1 and 5.0.2

UPDATED2: suspect PR
https://github.com/dotnet/fsharp/pull/11142

**Comment by Szer** [comment]
There is a suspect

https://github.com/dotnet/fsharp/pull/11142

**Comment by teo-tsirpanis** [comment]
@vzarytovskii should it be Area-Library?

**Comment by vzarytovskii** [comment]
> @vzarytovskii should it be Area-Library?

Yeah, that's right, misclicked.

**Comment by dsyme** [comment]
Fixed by #11799
